### PR TITLE
[Flutter-Parent] Add login landing screen and domain search screen

### PR DESCRIPTION
**Pull request by JordanMarshall**

### CODE REVIEW

**Comment by instructure-cx** [comment]
Scan submitted to Checkmarx

**Comment by instructure-cx** [comment]
### Checkmarx scan completed
[Full Scan Details](https://cxmgr.inssec.info/CxWebClient/ViewerMain.aspx?scanid=1010858&projectid=20)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|0
Low|0
Informational|0

#### Violation Summary
Severity|Count
---|---

File: apps/flutter_parent/lib/screens/domain_search/domain_search_screen.dart [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+              value: _loading ? null : 0,
+              backgroundColor: Colors.transparent,
+            ),
+          ),
+          Divider(
+            height: 0,
+            color: Colors.grey,
+          ),
+          Flexible(
+            flex: 10000,
```
Comment (CalvinKern) [inline apps/flutter_parent/lib/screens/domain_search/domain_search_screen.dart]: It's over 9000!

File: apps/flutter_parent/lib/screens/domain_search/domain_search_screen.dart [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+          Divider(
+            height: 0,
+            color: Colors.grey,
+          ),
+          Center(
+            child: FlatButton(
+              child: Text(AppLocalizations.of(context).domainSearchHelpLabel),
+              textTheme: ButtonTextTheme.accent,
+              onPressed: () {
+                _showHelpDialog(context);
```
Comment (CalvinKern) [inline apps/flutter_parent/lib/screens/domain_search/domain_search_screen.dart]: Missing tests for showing the dialog.

File: apps/flutter_parent/lib/screens/domain_search/domain_search_screen.dart [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+    for (var indexedSpan in indexedSpans) {
+      spans.add(TextSpan(text: text.substring(index, indexedSpan.key)));
+      spans.add(indexedSpan.value);
+      index = indexedSpan.key + indexedSpan.value.text.length;
+    }
+    spans.add(TextSpan(text: text.substring(index)));
+
+    return TextSpan(children: spans);
+  }
```
Comment (CalvinKern) [inline apps/flutter_parent/lib/screens/domain_search/domain_search_screen.dart]: This seems like a great helper function that could be extracted to a utility class.

File: apps/flutter_parent/test/screens/domain_search_page_test.dart [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+  testWidgets("Displays search results", (WidgetTester tester) async {
+    int count = 5;
+    MockInteractor interactor = MockInteractor()
+      ..schoolSearchResponse = (_) async {
+        return List.generate(
+          count,
+          (idx) => SchoolDomain((sd) => sd
+            ..domain = "test$idx.domains.com"
+            ..name = "Test domain $idx"),
```
Comment (CalvinKern) [inline apps/flutter_parent/test/screens/domain_search_page_test.dart]: I miss Kotlin, but cascading like this really is _super_ nice.

File: apps/flutter_parent/test/screens/domain_search_page_test.dart [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    await tester.pumpWidget(TestApp(DomainSearchScreen()));
+    await tester.pumpAndSettle();
+
+    await tester.enterText(find.byType(TextField), "one");
+    await tester.enterText(find.byType(TextField), "two");
+    await tester.enterText(find.byType(TextField), "three");
+    await tester.pumpAndSettle(Duration(milliseconds: 1000)); // Allow plenty of time for first async call to finish
+    expect(find.text("Query One"), findsNothing);
+    expect(find.text("Query Two"), findsNothing);
+    expect(find.text("Query Three"), findsOneWidget);
```
Comment (CalvinKern) [inline apps/flutter_parent/test/screens/domain_search_page_test.dart]: This is an amazing test! I very much like how we can test things like this so easily now!

File: apps/flutter_parent/lib/models/serializers.dart [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+/// If changes are made, run `flutter pub run build_runner build` from the project root. Alternatively, you can
+/// have it watch for changes and automatically build if you run `flutter pub run build_runner watch`.
+@SerializersFor([
+  SchoolDomain,
+])
+final Serializers serializers = _$serializers;
+
+Serializers jsonSerializers = (serializers.toBuilder()..addPlugin(StandardJsonPlugin())).build();
+
+T deserialize<T>(dynamic value) => jsonSerializers.deserializeWith<T>(serializers.serializerForType(T), value);
```
Comment (CalvinKern) [inline apps/flutter_parent/lib/models/serializers.dart]: Does this need to be `jsonSerializers.serializerForType(T)`?

File: apps/flutter_parent/lib/l10n/app_localizations.dart [hunk 7]
```diff
@@ -145,4 +145,62 @@ class AppLocalizations {
...
+        name: 'findSchoolOrDistrict',
+        desc: 'Text for the find-my-school button',
+      );
+
+  /// Domain search screen
+
+  String get domainSearchInputHint => Intl.message(
+        'Enter school name or district...',
+        name: 'domainSearchInputHint',
+        desc: "Input hint for the text box on the deomain search screen",
```
[the remark this answers is not in the document]
Comment (JordanMarshall) [reply apps/flutter_parent/lib/l10n/app_localizations.dart]: lol

File: apps/flutter_parent/test/screens/domain_search_screen_test.dart [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+    await tester.pumpAndSettle();
+
+    await tester.testTextInput.receiveAction(TextInputAction.done);
+
+    // TODO: Enable once WebLoginPage is implemented
+    //expect(find.byType(WebLoginPage), findsOneWidget);
+  });
+}
+
+class MockInteractor extends Mock implements DomainSearchInteractor {}
```
Comment (CalvinKern) [inline apps/flutter_parent/test/screens/domain_search_screen_test.dart]: I love how clean these tests can be when you just mock something like this.

**Pull request merged**